### PR TITLE
Migrate from protoc to rust-protobuf

### LINKED ISSUE #2645: Missing cmake
## Summary

For some packages like prost-build v0.10.3 requires cmake, and build failed.

```
error: failed to run custom build command for `prost-build v0.10.3`

Caused by:
  process didn't exit successfully: `./rust-libp2p/target/debug/build/prost-build-6fe0b99530fe295b/build-script-build` (exit status: 101)
  --- stdout
  cargo:rerun-if-changed=./.cargo/registry/src/github.com-1ecc6299db9ec823/prost-build-0.10.3/third-party/protobuf/cmake
  CMAKE_TOOLCHAIN_FILE_x86_64-apple-darwin = None
  CMAKE_TOOLCHAIN_FILE_x86_64_apple_darwin = None
  HOST_CMAKE_TOOLCHAIN_FILE = None
  CMAKE_TOOLCHAIN_FILE = None
  CMAKE_GENERATOR_x86_64-apple-darwin = None
  CMAKE_GENERATOR_x86_64_apple_darwin = None
  HOST_CMAKE_GENERATOR = None
  CMAKE_GENERATOR = None
  CMAKE_PREFIX_PATH_x86_64-apple-darwin = None
  CMAKE_PREFIX_PATH_x86_64_apple_darwin = None
  HOST_CMAKE_PREFIX_PATH = None
  CMAKE_PREFIX_PATH = None
  CMAKE_x86_64-apple-darwin = None
  CMAKE_x86_64_apple_darwin = None
  HOST_CMAKE = None
  CMAKE = None
  running: "cmake" "./.cargo/registry/src/github.com-1ecc6299db9ec823/prost-build-0.10.3/third-party/protobuf/cmake" "-DCMAKE_INSTALL_PREFIX=../rust-libp2p/target/debug/build/prost-build-1521adc0b256a667/out" "-DCMAKE_C_FLAGS= -ffunction-sections -fdata-sections -fPIC -m64 -arch x86_64" "-DCMAKE_C_COMPILER=/usr/bin/cc" "-DCMAKE_CXX_FLAGS= -ffunction-sections -fdata-sections -fPIC -m64 -arch x86_64" "-DCMAKE_CXX_COMPILER=/usr/bin/c++" "-DCMAKE_ASM_FLAGS= -ffunction-sections -fdata-sections -fPIC -m64 -arch x86_64" "-DCMAKE_ASM_COMPILER=/usr/bin/cc" "-DCMAKE_BUILD_TYPE=Debug"

  --- stderr
  thread 'main' panicked at '
  failed to execute command: No such file or directory (os error 2)
  is `cmake` not installed?

  build script failed, must exit now', .cargo/registry/src/github.com-1ecc6299db9ec823/cmake-0.1.48/src/lib.rs:975:5
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
warning: build failed, waiting for other jobs to finish...
error: build failed
```
## Actual behaviour

<details><summary>Debug Output</summary>
<p>

```
<output>
```
</p>
</details>

## Version

- libp2p version (version number, commit, or branch):
commit f04f6bb4fcf83a056b399dada4d57f844b29cd9a

## Would you like to work on fixing this bug?

Yes / No / Maybe.

### LINKED ISSUE #3024: Migrate from `protoc` to `quick-protobuf` and remove the buildscripts
## Description

Migrate our codebase from `protoc` to `quick-protobuf`.
We also want to get rid of the buildscripts and instead check the generated code into Git and add a CI step that verifies that they are directly generated from the protobuf files.

## Motivation

- Users are irritated about the dependencies required to build rust-libp2p: https://github.com/libp2p/rust-libp2p/issues/2645
- It causes bloat in our CI workflow: Every job needs to install protoc.
- We sometimes get CI failures in installing protoc: https://github.com/libp2p/rust-libp2p/actions/runs/3243659893/jobs/5327292307

## Current Implementation

We depend on protoc being installed because prost no longer ships with it.

## Are you planning to do it yourself in a pull request?

No.

### PR DESCRIPTION
**Pull request by kckeiks**
# Description

Closes #3024.
Fixes #2645.

## Links to any relevant issues

* https://github.com/stepancheg/rust-protobuf/pull/525

## Open Questions

I updated core and a couple of other packages. This is wip atm but I'll update the remaining crates this week.

## Change checklist

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: core/src/identity.rs
Comment (thomaseizinger) [inline core/src/identity.rs]: When can this fail?

File: core/build.rs
Comment (thomaseizinger) [inline core/build.rs]: It says [here](https://docs.rs/protobuf-codegen/latest/protobuf_codegen/struct.Customize.html#method.gen_mod_rs) that this is the default in version 3?

File: core/build.rs
Comment (thomaseizinger) [inline core/build.rs]: I think we don't use any of the features described in `lite_runtime`: https://docs.rs/protobuf-codegen/latest/protobuf_codegen/struct.Customize.html#method.lite_runtime

Can you check if we can enable that?

File: core/src/identity.rs
Comment (thomaseizinger) [reply core/src/identity.rs]: Looking at the docs a bit, I think this can only fail on bad usage and since that is static, we should be fine with panicking.

Would you mind investigating this in a bit more detail to confirm? Thank you :)

File: core/src/identity.rs
Comment (thomaseizinger) [inline core/src/identity.rs]: Are we roundtripping here? 🤔

File: core/src/peer_record.rs
Comment (thomaseizinger) [inline core/src/peer_record.rs:132]: Is this part of the public API? If yes, then this is actually a breaking change which is unfortunate :/

**Review by thomaseizinger — COMMENTED** [review]
Thank you for tackling this! Very appreciated!

I left a few questions.

File: core/src/peer_record.rs
Comment (kckeiks) [reply core/src/peer_record.rs:132]: Oh I missed that. Yes, it is part of the public API :disappointed:.

File: core/build.rs
Comment (kckeiks) [reply core/build.rs]: It wasn't working when I was testing it and I don't [see](https://docs.rs/protobuf-codegen/latest/src/protobuf_codegen/customize/mod.rs.html#163-166) how where it defaults to true. I'll test it again.

File: core/src/identity.rs
Comment (kckeiks) [reply core/src/identity.rs]: Now that I'm looking at it `keys_proto::KeyType::from_i32(pubkey.Type().value()` will not fail. ~I'll add an `expect`.~ I'll do something else :).

File: core/src/peer_record.rs
Comment (thomaseizinger) [reply core/src/peer_record.rs:132]: Damn okay. Well don't worry about it here for now, I'll send a PR that wraps the error here. That is still a breaking change but we can ship that separately!

File: core/src/identity.rs
Comment (kckeiks) [reply core/src/identity.rs]: Yeah I think we are good with panicking https://github.com/stepancheg/rust-protobuf/issues/404

File: core/build.rs
Comment (kckeiks) [reply core/build.rs]: You were right about this so I'm removing it. I must have overlooked something.

File: core/build.rs
Comment (kckeiks) [reply core/build.rs]: Done.

File: core/src/identity.rs
Comment (thomaseizinger) [reply core/src/identity.rs]: That is good to know but also annoying that we need to document it.

It would be nice if they'd implement a trait for all proto3 messages that has an infallible serializer.

Comment (thomaseizinger) [reply core/src/identity.rs]: That is good to know but also annoying that we need to document it.

It would be nice if they'd implement a trait for all proto3 messages that has an infallible serializer.

File: core/build.rs
Comment (thomaseizinger) [inline core/build.rs]: All the builder fns take `self` so I think we can just do this:

```suggestion
        .customize(protobuf_codegen::Customize::default().lite_runtime(true))
```

File: core/src/identity.rs
Comment (thomaseizinger) [inline core/src/identity.rs]: Can we say something like "all fields to be initialized" given that's the invariant we are relying on?

File: protocols/autonat/src/protocol.rs
Comment (thomaseizinger) [inline protocols/autonat/src/protocol.rs:303]: Well this is kind of an annoying API but I guess that is what you get with generated code.

File: protocols/dcutr/CHANGELOG.md
Comment (thomaseizinger) [inline protocols/dcutr/CHANGELOG.md:3]: We should mention in all changelogs that this removes the `protoc` dependency.

Comment (thomaseizinger) [reply protocols/dcutr/CHANGELOG.md:3]: Can you also remove the `protoc` install step from all workflows please?

File: transports/plaintext/src/error.rs
Comment (thomaseizinger) [inline transports/plaintext/src/error.rs:31]: This makes it a breaking change too I think.

**Review by thomaseizinger — COMMENTED** [review]
Thanks!

File: core/src/identity.rs
Comment (kckeiks) [reply core/src/identity.rs]: You beat me to it :blush: https://github.com/stepancheg/rust-protobuf/issues/404#issuecomment-1287460286.

File: core/src/peer_record.rs
Comment (thomaseizinger) [reply core/src/peer_record.rs:132]: So my idea here is to make a dedicated error like this:

```rust
#[derive(thiserror::Error, Debug)]
#[error(transparent)]
pub struct DecodeError(protobuf::Error)
```

This preserves the source chain but hides the concrete error type from the public API.

It'd be nice to do this in a separate PR and then put this feature on top of that. Let me know if you'd be happy to do that otherwise I'll try to get to that next week!

File: transports/plaintext/src/error.rs
Comment (kckeiks) [reply transports/plaintext/src/error.rs:31]: Yes, it's public.

**Comment by thomaseizinger** [comment]
I kicked CI, seems there are still a couple of compile errors 😊

File: protocols/autonat/src/protocol.rs
Comment (kckeiks) [reply protocols/autonat/src/protocol.rs:303]: > Well this is kind of an annoying API

Agreed. It would be nice if at least it generated setters for `MessageField`s.

**Comment by kckeiks** [comment]
> I kicked CI, seems there are still a couple of compile errors blush

Yes, I need to clean up a few tests. Will fix.

File: core/src/peer_record.rs
Comment (kckeiks) [reply core/src/peer_record.rs:132]: > So my idea here is to make a dedicated error like this:
> 
> ```rust
> #[derive(thiserror::Error, Debug)]
> #[error(transparent)]
> pub struct DecodeError(protobuf::Error)
> ```
> 
> This preserves the source chain but hides the concrete error type from the public API.

That sounds good! 

> It'd be nice to do this in a separate PR and then put this feature on top of that. Let me know if you'd be happy to do that otherwise I'll try to get to that next week!

If you could please do that that would be great! I'll rebase once that is merged.

File: misc/prost-codec/Cargo.toml
Comment (thomaseizinger) [inline misc/prost-codec/Cargo.toml:16]: This isn't really a `prost-codec` anymore, now that we are replacing the implementation with `protobuf` :sweat_smile: 

@mxinden Should we publish a new crate?

File: core/src/peer_record.rs
Comment (thomaseizinger) [reply core/src/peer_record.rs:132]: I starting working on this here: https://github.com/libp2p/rust-libp2p/pull/3058

**Comment by kckeiks** [comment]
I think there is one crate left to do. I should have time to do it tomorrow and also take care of some of the other pending tasks.

File: protocols/relay/src/v2/protocol/inbound_hop.rs
Comment (thomaseizinger) [inline protocols/relay/src/v2/protocol/inbound_hop.rs]: Can we retain the order here please (`status` being set last)? Makes the diff easier to read :)

File: protocols/rendezvous/src/codec.rs
Comment (thomaseizinger) [inline protocols/rendezvous/src/codec.rs]: This can be solved with the `ok_or` combinator on `Option`.

File: protocols/rendezvous/src/codec.rs
Comment (thomaseizinger) [inline protocols/rendezvous/src/codec.rs]: I think this may be cleaner too if we'd use `ok_or(ConversionError::InconsistentWireMessage)?` here and unpack the fields one by one instead of pattern matching on it.

File: protocols/rendezvous/src/codec.rs
Comment (thomaseizinger) [inline protocols/rendezvous/src/codec.rs]: This `status` is enum as well right? I think we should probably use that instead of the bare number. With `prost`, we couldn't match on that but now we can :)

**Review by thomaseizinger — COMMENTED** [review]
Only a few minor comments!

Looking forward to the full solution :)

File: protocols/rendezvous/src/codec.rs
Comment (kckeiks) [reply protocols/rendezvous/src/codec.rs]: It's an enum but it's wrapped in a couple of abstractions. Will take a look again.

File: misc/prost-codec/Cargo.toml
Comment (mxinden) [reply misc/prost-codec/Cargo.toml:16]: :+1:

**Comment by mxinden** [comment]
First off, thanks @kckeiks for the work here. This will make using rust-libp2p for newcomers so much easier.

Given that Protobuf is at the core of libp2p, performance is relevant. Are there any comparisons to prost you can point me to @thomaseizinger @kckeiks? Let's make sure to test this, e.g. via a large user, before merging here.

**Comment by kckeiks** [comment]
> First off, thanks @kckeiks for the work here. This will make using rust-libp2p for newcomers so much easier.

:+1: 

> Given that Protobuf is at the core of libp2p, performance is relevant. Are there any comparisons to prost you can point me to @thomaseizinger @kckeiks? Let's make sure to test this, e.g. via a large user, before merging here.

[quick-proto](https://github.com/tafia/quick-protobuf/tree/master/perftest) has some performance comparisons of `prost`, `protobuf` and `quick-protobuf`. `protobuf` looks slower than the rest. However, these results are from 4 years ago.

**Comment by kckeiks** [comment]
> I kicked CI, seems there are still a couple of compile errors blush

I fixed the tests.

**Comment by jxs** [comment]
`quick-proto` [perfest](https://github.com/tafia/quick-protobuf/tree/master/perftest) with `protobuf` `3.2.0`  and `prost` `0.11` (pushed the changes [here](https://github.com/jxs/rust-protobuf/tree/update-perftest))
```
         labels   rust-protobuf  quick-protobuf           prost      quick/rust      prost/rust
                        ns/iter         ns/iter         ns/iter               %               %

test1
          write             632             471             470            25.5            25.6
           read             966             448             539            53.6            44.2
    read no vec             853             330             440            61.3            48.4
     read reuse             788                              NA              NA

test_repeated_bool
          write             859            1010             890           -17.6            -3.6
           read            1617             924            1136            42.9            29.7
    read no vec            1519             902            1082            40.6            28.8
     read reuse            1027                              NA              NA

test_repeated_packed_int32
          write            1736            1753            1595            -1.0             8.1
           read            2290            1693            1655            26.1            27.7
    read no vec            2261            1622            1575            28.3            30.3
     read reuse            1708                              NA              NA

test_repeated_messages
          write            4542            6754            7764           -48.7           -70.9
           read            9943            6884            7047            30.8            29.1
    read no vec            6616            3939            4032            40.5            39.1
     read reuse            5802                              NA              NA

test_optional_messages
          write            2062            1702            1768            17.5            14.3
           read            4420            1768            2322            60.0            47.5
    read no vec            4390            1949            2626            55.6            40.2
     read reuse            4328                              NA              NA

test_strings
          write            1141             970             951            15.0            16.7
           read            3114             883            2772            71.6            11.0
    read no vec            2972             695            2727            76.6             8.2
     read reuse            2841                              NA              NA

test_small_bytearrays
          write             992             890             900            10.3             9.3
           read            1860             652            2651            64.9           -42.5
    read no vec            1783             551            2445            69.1           -37.1
     read reuse            1638                              NA              NA

test_large_bytearrays
          write           22729           14337           25541            36.9           -12.4
           read           23667            4691           27981            80.2           -18.2
    read no vec            7103             687            8872            90.3           -24.9
     read reuse            5393                              NA              NA
Also (not performance related but fun to know) - Rpc test was successful!
```

**Comment by kckeiks** [comment]
Thank you for running those tests @jxs !

File: core/src/identity.rs
Comment (thomaseizinger) [inline core/src/identity.rs]: We discovered that this can never fail right? Let's panic here for consistency.

**Comment by thomaseizinger** [comment]
@kckeiks Please decide if you want to amend commits and force-push or never force-push at all :)

We squash merge in the end so the commits will be gone anyway. But reviews can sometimes be easier by amending commits later but that is also not happening here: For example 8acaa31ce2b9bedb034a9b473625540bbfa8c3c6 would need to be squashed into ea8a3038a7ac9d49ac89ff996ea74f14ef447a87.

Force-pushing but not amending commits is kind of the worst of both worlds: I can't see a clear diff from the last push but the commits also overlap, making it hard to review patch-by-patch :)

File: core/src/identity.rs
Comment (thomaseizinger) [inline core/src/identity.rs:180]: Please panic here for consistency.

File: core/src/identity.rs
Comment (thomaseizinger) [inline core/src/identity.rs:195]: This looks like another round-trip to me? Wouldn't `Type()` already give you the enum value?

File: core/src/peer_record.rs
Comment (thomaseizinger) [inline core/src/peer_record.rs:38]: Why do we need the `map_err` now?

File: protocols/gossipsub/src/behaviour.rs
Comment (thomaseizinger) [inline protocols/gossipsub/src/behaviour.rs:744]: If we panic here, we might as well also panic on line 621.

Comment (thomaseizinger) [reply protocols/gossipsub/src/behaviour.rs:744]: I think as regular cast could do as well.

**Review by thomaseizinger — CHANGES_REQUESTED** [review]
Thanks!

A few more comments, plus the CI workflow needs adjusting to remove the installation of `protoc`!

**Comment by thomaseizinger** [comment]
> `quick-proto` [perfest](https://github.com/tafia/quick-protobuf/tree/master/perftest?rgh-link-date=2022-10-25T22%3A24%3A39Z) with `protobuf` `3.2.0` and `prost` `0.11` (pushed the changes [here](https://github.com/jxs/rust-protobuf/tree/update-perftest?rgh-link-date=2022-10-25T22%3A24%3A39Z))
> 
> ```
>          labels   rust-protobuf  quick-protobuf           prost      quick/rust      prost/rust
>                         ns/iter         ns/iter         ns/iter               %               %
> 
> test1
>           write             632             471             470            25.5            25.6
>            read             966             448             539            53.6            44.2
>     read no vec             853             330             440            61.3            48.4
>      read reuse             788                              NA              NA
> 
> test_repeated_bool
>           write             859            1010             890           -17.6            -3.6
>            read            1617             924            1136            42.9            29.7
>     read no vec            1519             902            1082            40.6            28.8
>      read reuse            1027                              NA              NA
> 
> test_repeated_packed_int32
>           write            1736            1753            1595            -1.0             8.1
>            read            2290            1693            1655            26.1            27.7
>     read no vec            2261            1622            1575            28.3            30.3
>      read reuse            1708                              NA              NA
> 
> test_repeated_messages
>           write            4542            6754            7764           -48.7           -70.9
>            read            9943            6884            7047            30.8            29.1
>     read no vec            6616            3939            4032            40.5            39.1
>      read reuse            5802                              NA              NA
> 
> test_optional_messages
>           write            2062            1702            1768            17.5            14.3
>            read            4420            1768            2322            60.0            47.5
>     read no vec            4390            1949            2626            55.6            40.2
>      read reuse            4328                              NA              NA
> 
> test_strings
>           write            1141             970             951            15.0            16.7
>            read            3114             883            2772            71.6            11.0
>     read no vec            2972             695            2727            76.6             8.2
>      read reuse            2841                              NA              NA
> 
> test_small_bytearrays
>           write             992             890             900            10.3             9.3
>            read            1860             652            2651            64.9           -42.5
>     read no vec            1783             551            2445            69.1           -37.1
>      read reuse            1638                              NA              NA
> 
> test_large_bytearrays
>           write           22729           14337           25541            36.9           -12.4
>            read           23667            4691           27981            80.2           -18.2
>     read no vec            7103             687            8872            90.3           -24.9
>      read reuse            5393                              NA              NA
> Also (not performance related but fun to know) - Rpc test was successful!
> ```

Those relative numbers aren't exactly great. I'd be willing to take that hit though for the removal of the `protoc` dependency.

**Comment by kckeiks** [comment]
> @kckeiks Please decide if you want to amend commits and force-push or never force-push at all :)
> 
> We squash merge in the end so the commits will be gone anyway. But reviews can sometimes be easier by amending commits later but that is also not happening here: For example [8acaa31](https://github.com/libp2p/rust-libp2p/commit/8acaa31ce2b9bedb034a9b473625540bbfa8c3c6) would need to be squashed into [ea8a303](https://github.com/libp2p/rust-libp2p/commit/ea8a3038a7ac9d49ac89ff996ea74f14ef447a87).
> 
> Force-pushing but not amending commits is kind of the worst of both worlds: I can't see a clear diff from the last push but the commits also overlap, making it hard to review patch-by-patch :)

Sorry, I've been relying on my command history too much while amending and force-pushing so I inadvertently force-pushed. I will cut that out :smile_cat: .

**Comment by kckeiks** [comment]
Any thoughts on using `quick-protobuf` instead? Here is a list of some of its missing features https://github.com/tafia/quick-protobuf/issues/12. I tried it out and the API seems simpler than `protobuf`. There hasn't been much recent activity/development though.

**Comment by thomaseizinger** [comment]
> Any thoughts on using `quick-protobuf` instead? Here is a list of some of its missing features [tafia/quick-protobuf#12](https://github.com/tafia/quick-protobuf/issues/12). I tried it out and the API seems simpler than `protobuf`. There hasn't been much recent activity/development though.

Can you share an example of what `quick-protobuf` looks like if we were to integrate it into our codebase? I don't mind much which implementation we are using, as long as we get rid of the `protoc` dependency :)

**Comment by jxs** [comment]
> Those relative numbers aren't exactly great. I'd be willing to take that hit though for the removal of the `protoc` dependency.

yeah agree Thomas.

>Any thoughts on using quick-protobuf instead? Here is a list of some of its missing features https://github.com/tafia/quick-protobuf/issues/12. I tried it out and the API seems simpler than protobuf. There hasn't been much recent activity/development though.

`rust-protobuf` [also seems to be looking for maintainers](https://github.com/stepancheg/rust-protobuf#help)

**Comment by kckeiks** [comment]
> > Any thoughts on using `quick-protobuf` instead? Here is a list of some of its missing features [tafia/quick-protobuf#12](https://github.com/tafia/quick-protobuf/issues/12). I tried it out and the API seems simpler than `protobuf`. There hasn't been much recent activity/development though.
> 
> Can you share an example of what `quick-protobuf` looks like if we were to integrate it into our codebase? I don't mind much which implementation we are using, as long as we get rid of the `protoc` dependency :)

This PR https://github.com/libp2p/rust-libp2p/pull/3066 uses `quick-proto` in `libp2p-core`.

**Pull request closed without merging**
